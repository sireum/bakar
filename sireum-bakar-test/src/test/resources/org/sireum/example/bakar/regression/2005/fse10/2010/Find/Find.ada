package Find_Package
is
   type Index is range 1..10;
   type AType is array (Index) of Integer;
   -- subtype Index is Integer range 1 .. 10;
   -- type Table is array(Index) of Integer;

   function Find(A : AType; X : Integer) return Index;
   --# pre for some I in Index => ( A(I) = X );
   --# return Z => (A(Z) = X and
   --#    (for all I in Index range Index'First .. (Z-1) =>
   --#        (A(I) /= X)));

end Find_Package;

package body Find_Package
is
   function Find (A: AType; X : Integer) return Index
   is
      Z : Index := Index'First;
   begin
      while A(Z) /= X and Z < Index'Last loop
         --# assert Z < Index'Last
         --#    and (for all  I in Index range Index'First .. Z  => (A(I) /= X))
         --#    and (for some I in Index range Z+1 .. Index'Last => (A(I)  = X ));
         --  Replacing Z+1 by Index'First in the last conjunct results in
         --  fewer VCs remaining undischarged by the Spark tools, but IMHO, it
         --  yields VCs that should be easier to prove manually.
         Z := Z + 1;
      end loop;
      return Z;
   end Find;

end Find_Package;
