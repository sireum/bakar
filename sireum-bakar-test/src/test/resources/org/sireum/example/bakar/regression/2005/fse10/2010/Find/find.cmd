3.
delete [2, 3, 4, 6, 7].
list
unwrap h # 1.
unwrap c # 1.
list
done
list
simp h # 11.
list.
done.
instantiate int_I__1 with int_i__1.
done.
list.
help.
help simplify.
simplify.
list.
delete [1, 5, 8, 9, 10, 12, 13].
done.
list.
