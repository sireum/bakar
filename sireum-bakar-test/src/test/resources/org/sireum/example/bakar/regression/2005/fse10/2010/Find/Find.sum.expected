-------------------------------------------------------------------------------
                          Semantic Analysis Summary                            
                                POGS GPL 2010                                  
            Copyright (C) 2010 Altran Praxis Limited, Bath, U.K.               
-------------------------------------------------------------------------------

Summary of:

Verification Condition files (.vcg)
Simplified Verification Condition files (.siv)
Proof Logs (.plg)
Dead Path Conjecture files (.dpc)
Summary Dead Path files (.sdp)

"status" column keys:
    1st character:
        '-' - No VC
        'S' - No SIV
        'U' - Undischarged
        'E' - Proved by Examiner
        'I' - Proved by Simplifier by Inference
        'X' - Proved by Simplifier by Contradiction
        'P' - Proved by Simplifier using User Defined Proof Rules
        'C' - Proved by Checker
        'R' - Proved by Review
        'F' - VC is False
    2nd character:
        '-' - No DPC
        'S' - No SDP
        'U' - Unchecked
        'D' - Dead path
        'L' - Live path

in the directory:
/Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Find

Summary produced: 14-SEP-2010 13:43:36.17

File /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Find/p/find.vcg
function P.Find

VCs generated 14-SEP-2010 13:43:35

VCs simplified 14-SEP-2010 13:43:35

VCs for function_find :
 -----------------------------------------------------------------------------
| #   | From  | To                  | Proved By          | Dead Path | Status |
|-----------------------------------------------------------------------------
| 1   | start | rtc check @ 22      | Inference          | No DPC    |   I-   |
| 2   | 23    | rtc check @ 22      | Inference          | No DPC    |   I-   |
| 3   | start |    assert @ 23      | Undischarged       | No DPC    |   U-   |
| 4   | 23    |    assert @ 23      | Undischarged       | No DPC    |   U-   |
| 5   | 23    | rtc check @ 29      | Inference          | No DPC    |   I-   |
| 6   | start |    assert @ finish  | Inference          | No DPC    |   I-   |
| 7   | 23    |    assert @ finish  | Undischarged       | No DPC    |   U-   |
 -----------------------------------------------------------------------------


===============================================================================
Summary:

The following subprograms have undischarged VCs (excluding those proved false):

   3  /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Find/p/find.vcg

Proof strategies used by subprograms
-------------------------------------------------------------------------
Total subprograms with at least one VC proved by examiner:              0
Total subprograms with at least one VC proved by simplifier:            1
Total subprograms with at least one VC proved by contradiction:         0
Total subprograms with at least one VC proved with user proof rule:     0
Total subprograms with at least one VC proved using checker:            0
Total subprograms with at least one VC discharged by review:            0

Maximum extent of strategies used for fully proved subprograms:
-------------------------------------------------------------------------
Total subprograms with proof completed by examiner:                     0
Total subprograms with proof completed by simplifier:                   0
Total subprograms with proof completed with user defined rules:         0
Total subprograms with proof completed by checker:                      0
Total subprograms with VCs discharged by review:                        0

Overall subprogram summary:
-------------------------------------------------------------------------
Total subprograms fully proved:                                         0
Total subprograms with at least one undischarged VC:                    1  <<<
Total subprograms with at least one false VC:                           0
                                                                    -----
Total subprograms for which VCs have been generated:                    1


ZombieScope Summary:
-------------------------------------------------------------------------
Total subprograms for which DPCs have been generated:                   0
Total number subprograms with dead paths found:                         0
Total number of dead paths found:                                       0


VC summary:
-------------------------------------------------------------------------
Note: U/R denotes where the Simplifier has proved VCs using one or more user-
defined proof rules.

Total VCs by type:                                       
                            -----------Proved By Or Using------------
                     Total  Examiner Simp(U/R)  Checker Review False Undiscgd
Assert or Post:          4      0      1              0      0      0       3
Precondition check:      0      0      0              0      0      0       0
Check statement:         0      0      0              0      0      0       0
Runtime check:           3      0      3              0      0      0       0
Refinement VCs:          0      0      0              0      0      0       0
Inheritance VCs:         0      0      0              0      0      0       0
===============================================================================
Totals:                  7      0      4              0      0      0       3  <<<
% Totals:                       0%    57%             0%     0%     0%     43% <<<
===================== End of Semantic Analysis Summary ========================
